torchaudio/models/wav2vec2/components.py

```diff
@@ -0,0 +1,689 @@
+
+    def forward(
+            self,
+            x: Tensor,
+            attention_mask: Optional[Tensor] = None,
+    ):
+        """
+        Args:
+            x (Tensor): shape: ``(batch, sequence_length, embed_dim)``
+            attention_mask (Tensor, optional):
+                shape: ``(batch, 1, sequence_length, sequence_length)``
+        """
+        residual = x
+
+        if self.layer_norm_first:
+            x = self.layer_norm(x)
+
+        x = self.attention(x, attention_mask)
+        x = self.dropout(x)
+        x = residual + x
+
+        if self.layer_norm_first:
+            x = x + self.feed_forward(self.final_layer_norm(x))
+        else:
+            x = self.layer_norm(x)
+            x = self.final_layer_norm(x + self.feed_forward(x))
+        return x
+
+
+class Transformer(Module):
+    def __init__(
+            self,
+            pos_conv_embed: Module,
+            dropout: float,
+            layers: Module,
+            layer_norm_first: bool,
+            layer_drop: float,
+    ):
+        super().__init__()
+        self.pos_conv_embed = pos_conv_embed
+        self.layer_norm = nn.LayerNorm(pos_conv_embed.embed_dim)
+        self.layer_norm_first = layer_norm_first
+        self.layer_drop = layer_drop
+        self.dropout = nn.Dropout(dropout)
+        self.layers = layers
+
+    def forward(
+            self,
+            x: Tensor,
+            attention_mask: Optional[Tensor] = None,
+    ):
+        x = x + self.pos_conv_embed(x)
+
+        if self.layer_norm_first:
+            x = self.layer_norm(x)
+
+        x = self.dropout(x)
+        for layer in self.layers:
+            # It is more readable if the following condition is flipped and we use `continue`,
+            # but TorchScript does not support `continue`.
+            if not self.training or torch.rand(1).item() > self.layer_drop:
+                x = layer(x, attention_mask)
+
+        if not self.layer_norm_first:
+            x = self.layer_norm(x)
+
+        return x
+
+
+class Encoder(Module):
+    def __init__(
+            self,
+            feature_projection: Module,
+            transformer: Module,
+            readout: Module,
+    ):
+        super().__init__()
+        self.feature_projection = feature_projection
+        self.transformer = transformer
+        self.readout = readout
+
+    def forward(
+            self,
+            features: Tensor,
+            lengths: Optional[Tensor] = None,
+    ) -> Tensor:
+        x = self.feature_projection(features)
+
+        mask: Optional[Tensor] = None
+        if lengths is not None:
+            batch_size, max_len, _ = x.shape
+            # create mask for padded elements and zero-out them
+            mask = torch.arange(max_len).expand(batch_size, max_len) >= lengths[:, None]
+            x[mask] = 0.0
+            # extend the mask to attention shape and set weight
+            mask = -10000.0 * mask[:, None, None, :].to(dtype=features.dtype)
```

Comment:
There's contention around what the right value here is. Parlai uses [neginf](https://github.com/facebookresearch/ParlAI/blob/2f48dcd9ab7b44b09001807b6a8a5e783972ad59/parlai/utils/torch.py#L37) that depends on the input dtype, nn.MHA uses [float(-inf)](https://github.com/pytorch/pytorch/blob/032d6b0643f10fdfd5bcc1249d045f86c4ecc454/torch/nn/functional.py#L5074) which likely has issues with lower precision dtypes and FasterTransformer uses [-10000](https://github.com/NVIDIA/FasterTransformer/blob/main/fastertransformer/cuda/attention_kernels.cu#L423) as well, fairseq uses [float("-inf")](https://github.com/pytorch/fairseq/blob/207254bf56374831d08c20064ca7a2740a871ce5/fairseq/modules/multihead_attention.py#L353) as well. I think the -10000.0 is fine, but I'm curious about your reasoning behind this.

Reply:
This value is from [HuggingFaces's implementation of Wav2Vec2.0](https://github.com/huggingface/transformers/blob/b3429ab678d0c782b0a1cc6df76f848b5c013f91/src/transformers/models/wav2vec2/modeling_wav2vec2.py#L533).

Let me try `float("-inf")` and if the test passes, then I will switch to `float("-inf")`.

Reply:
I cannot make it work with `float("-inf")`, so I will stick with `-10000`.